metrics/run/collect_metrics.go

```diff
@@ -138,89 +143,199 @@ func main() {
-		*outputGcsBucket)
+	log.Println("Uploading metrics")
 
 	outputBucket := gcsClient.Bucket(*outputGcsBucket)
-	outputCtx := storage.Context{
-		Context: ctx,
-		Client:  gcsClient,
-		Bucket: &storage.Bucket{
-			Name:   outputGcsBucket,
-			Handle: outputBucket,
+	datastoreClient, err := datastore.NewClient(ctx, *projectId)
+	if err != nil {
+		log.Fatal(err)
+	}
+	bigqueryClient, err := bigquery.NewClient(ctx, *projectId)
+	if err != nil {
+		log.Fatal(err)
+	}
+	outputters := [2]storage.Outputter{
+		storage.GCSDatastoreContext{
+			Context: ctx,
+			Bucket: storage.Bucket{
+				Name:   *outputGcsBucket,
+				Handle: outputBucket,
+			},
+			Client: datastoreClient,
+		},
+		storage.BQContext{
+			Context: ctx,
+			Client:  bigqueryClient,
 		},
 	}
-	metricsRun := metrics.MetricsRun{
-		StartTime: &readStartTime,
-		EndTime:   &readEndTime,
-		TestRuns:  &runs,
+
+	gcsDir := fmt.Sprintf("%d-%d", readStartTime.Unix(),
+		readEndTime.Unix())
+	passRatesBasename := "pass-rates"
+	passRateGCSPath := fmt.Sprintf("%s/%s.json.gz", gcsDir,
+		passRatesBasename)
+	passRatesUrl := fmt.Sprintf(
+		"https://storage.cloud.google.com/%s/%s",
+		*outputGcsBucket,
+		passRateGCSPath)
+	failuresBasenamef := func(browserName string) string {
+		return fmt.Sprintf("%s-failures", browserName)
+	}
+	failuresGCSPathf := func(browserName string) string {
+		return fmt.Sprintf("%s/%s.json.gz", gcsDir,
+			failuresBasenamef(browserName))
+	}
+	failuresUrlf := func(browserName string) string {
+		return fmt.Sprintf(
+			"https://storage.cloud.google.com/%s/%s",
+			*outputGcsBucket,
+			failuresGCSPathf(browserName))
+	}
+	passRateMetadata := metrics.PassRateMetadata{
+		StartTime: readStartTime,
+		EndTime:   readEndTime,
+		TestRuns:  runs,
+		DataUrl:   passRatesUrl,
 	}
 
-	outputDir := fmt.Sprintf("%d-%d", metricsRun.StartTime.Unix(),
-		metricsRun.EndTime.Unix())
-	outputErrs := make(chan error)
-	log.Printf("Writing to bucket directory %s\n", outputDir)
-	wg.Add(2 + len(failuresMetrics))
-	go func() {
-		defer wg.Done()
-		objName := fmt.Sprintf("%s/pass-rates.json.gz", outputDir)
-		passRateSummary := metrics.MetricsRunData{
-			MetricsRun: &metricsRun,
-			Data:       &passRateMetric,
-		}
-		err := storage.UploadMetricsRunData(&outputCtx, &objName,
-			&passRateSummary)
-		if err != nil {
-			log.Println(err)
-			outputErrs <- err
-		}
-	}()
-	go func() {
-		defer wg.Done()
-		objName := fmt.Sprintf("%d-%d/test-counts.json.gz",
-			metricsRun.StartTime.Unix(),
-			metricsRun.EndTime.Unix())
-		totalsSummary := metrics.MetricsRunData{
-			MetricsRun: &metricsRun,
-			Data:       &totals,
+	wg.Add((1 + len(failuresMetrics)) * len(outputters))
+	processUploadErrors := func(errs []error) {
+		for _, err := range errs {
+			log.Printf("Upload error: %v", err)
 		}
-		err := storage.UploadMetricsRunData(&outputCtx, &objName,
-			&totalsSummary)
-		if err != nil {
-			log.Println(err)
-			outputErrs <- err
+		if len(errs) > 0 {
+			log.Fatal(err)
```

Comment:
You've fallen out of the loop here, so err is back to the function-scoped var?

Reply:
You're right. This should be `errs[len(errs)-1]` to get the most recent of the upload errors.